feat: DynamicAnnotation project - FE

artikandri opened the pull request:
Features added with this PR: 
- Ability to create the `Dynamic Annotation` project from FE side
- Ability to view the dimensions in the `Dynamic Annotation` project
- Ability to create dimensions in  the `Dynamic Annotation` project
- Ability to assign predefined dimensions in  the `Dynamic Annotation` project
- Ability to annotate a text with given dimension in the `Dynamic Annotation` project
- Translations of some texts 

To test this branch, rebase the branch with the `-be` and `-api` branch first; or alternatively, test the  `-fe` branch instead (should be identical except for that branch is rebase continously with the other branches; this branch is solely made for the PR)

Notable limitations: 
- Currently the `min` and `max` fields in the `create dimensions` form only receive integer values. Additionally, giving uncommon min/max/step values to the `v-slider` will create CSS bugs in which the `slider ticks` are not in line with the `slider tick labels`. Fixing this will take time because the `v-slider` component belongs to the `vuetify` library, so I decided to leave it as it is. 
- In the slider input, when users annotate a dimension by clicking on the slider, then clicking on the `checkbox` and then unclick it again, then the previous slider value is not saved and the value will be reset to 0. This is because some errors in which the Vue component doesnt save the `tempValue`. I wasnt able to fix it and I decided to leave it as it is. 
- I didn't test the statistics page, but I think it should have no visible changes since the project was largely based on the `affective annotation` project
- Currently all predefined json labels are automatically uploaded in the `creation` step with no regards to the selected dimensions; this is due to FE's inability to check the JSON content and filter the content based on selected dimensions. 

The connection to the API was largely based on the approach made by Teddy in https://github.com/CLARIN-PL/doccano/pull/95 

What's new: 
- `frontend/assets/images/tasks/dynamic-annotation.png`: dynamic annotation project thumbnail image 
- `frontend/components/label/SliderForm.vue`: slider form, loaded in the `create new dimension` page 
- `frontend/components/label/CheckboxForm.vue`: checkbox form, loaded in the `create new dimension` page 
- `frontend/components/label/DimensionList.vue`: dimension list table, loaded in the `labels` page (dimension tab)
- `frontend/components/label/FormCreateDimension.vue`: create dimension form component 
- `frontend/components/tasks/dynamicAnnotation/CheckboxInput.vue`: dynamic checkbox input 
- `frontend/components/tasks/dynamicAnnotation/DimensionInput.vue`: dimension input 
- `frontend/components/tasks/dynamicAnnotation/SliderInput.vue`: dynamic slider input 
- `frontend/domain/models/dimension/dimensionRepository.ts`: dynamic dimension repository made by Teddy, won't be included in this PR merge if Teddy merge his branch first 
- `frontend/repositories/dimension/apiDimensionRepository.ts`: dynamic dimension repository made by Teddy, won't be included in this PR merge if Teddy merge his branch first 
- `frontend/services/application/dimension/dimensionApplicationServices.ts`: dynamic dimension repository made by Teddy, won't be included in this PR merge if Teddy merge his branch first 
- `frontend/services/application/dimension/dimensionData.ts`: dynamic dimension data for Typescript 
- `frontend/pages/projects/_id/dynamic-annotation/index.vue`: dynamic annotation annotation page
- `frontend/utils/stringHelpers.js`: new helpers for strings
- `frontend/utils/numberHelpers.js`: new helpers for numbers

What's changed: 
- `frontend/components/label/FormCreate.vue`: required rules translation
- `frontend/components/label/LabelList.vue`: minor bugs
- `frontend/components/layout/TheSideBar.vue`: exclude `DynamicAnnotation` project
- `frontend/components/member/FormCreate.vue`: required rules translation
- `frontend/components/project/FormCreate.vue`: handle `dynamic annotation` project creation
- `frontend/components/tasks/affectiveAnnotation/inputs/TextfieldModal.vue`: allowed number
- `frontend/domain/models/project/project.ts`: adjustments to allow `DynamicAnnotation` project creation
- `frontend/i18n/*/generic.js`: translations
- frontend/i18n/*/rules.js: translations
- frontend/i18n/*/labels.js
- `frontend/middleware/check-admin.js`: based on Teddy's approach but instead of creating 2 try and catches, I merged them into 1 block 
- `frontend/pages/projects/_id/dataset/index.vue`: handled `dynamic annotation` project dataset, based on `affective annotation`
- `frontend/pages/projects/_id/labels/_label_id/edit.vue`: grouped the functions into 1 function
- `frontend/pages/projects/_id/labels/add.vue`: handled `dimension creation` for dynamic annotation project
- `frontend/pages/projects/_id/labels/index.vue`: handled dimension tab in dynamic annotation project
- `frontend/pages/projects/create.vue`: handled `dynamic annotation` project creation
- `frontend/pages/statistics/index.vue`: required rules translation 
- `frontend/repositories/project/apiProjectRepository.ts`: minor adjustments
- `frontend/services/application/label/labelData.ts`: handling labelData in typescript 
- `frontend/services/application/project/projectApplicationService.ts`: dimension creation 
- `frontend/services/application/project/projectData.ts`: dimension creation 

Screenshots: 
![image](https://user-images.githubusercontent.com/12537724/226664839-3a7c8fcd-bcb9-478d-bdc5-19eedda44986.png)
![image](https://user-images.githubusercontent.com/12537724/226664939-f074a2b7-8154-48d4-9eed-0abae3df1e57.png)
![image](https://user-images.githubusercontent.com/12537724/226665075-ad2ba066-1030-4dea-89ee-37b8e23c4ac2.png)
![image](https://user-images.githubusercontent.com/12537724/226665687-da307ed5-eeee-4cba-ba73-b476c3f66e27.png)
![image](https://user-images.githubusercontent.com/12537724/226665759-956f8c7e-840a-4052-8247-4df63d90555b.png)
![image](https://user-images.githubusercontent.com/12537724/226665832-8f717ca5-0e7b-45ce-87c4-e6156c31d6c6.png)
![image](https://user-images.githubusercontent.com/12537724/226665915-3ee2287b-a0f9-4e04-92e4-36be6ed1c060.png)
![image](https://user-images.githubusercontent.com/12537724/226665997-2ef491ab-77ee-408c-8a6f-ad91f3dffd77.png)
![image](https://user-images.githubusercontent.com/12537724/226666046-3a2f00c0-2d89-4e8b-a803-7866caab5753.png)
![image](https://user-images.githubusercontent.com/12537724/226666090-7c8493f2-fc8f-483e-b4e0-0c239ad452b2.png)
![image](https://user-images.githubusercontent.com/12537724/226666140-b7ab7ace-a84a-4c73-9a42-77f9b9f79c3d.png)
![image](https://user-images.githubusercontent.com/12537724/226666224-10fcf79a-ff60-4350-90d9-1e44ee21da54.png)
![image](https://user-images.githubusercontent.com/12537724/226667083-ae8d5954-ca95-4d28-b014-4b255e0b81f7.png)
![image](https://user-images.githubusercontent.com/12537724/226667235-0f4199d8-a80d-4c62-93a2-05d1e4a914f3.png)
![image](https://user-images.githubusercontent.com/12537724/226667330-a11a10d8-4bb4-4432-81d6-b21b4d77d15f.png)
![image](https://user-images.githubusercontent.com/12537724/226667403-4d19ddd6-14c7-4e9e-ae22-3b135745a727.png)
![image](https://user-images.githubusercontent.com/12537724/226667575-6a115c3f-a81f-4452-8b5a-f5504b654b72.png)
![image](https://user-images.githubusercontent.com/12537724/226667662-b569f076-e67a-42cb-990d-9ad472eb092e.png)


Edit: 
I forgot to mention that now the min-max fields for the slider is 1-20 and the max difference for them is 20.

teddy-f-47 commented on frontend/pages/projects/_id/labels/index.vue:
```diff
@@ -173,9 +205,40 @@ export default Vue.extend({
+      if (items) {
+        this.items = _.cloneDeep(items)
+        this.isDimensionsTab && this.setCurrentDimensions(items)
+      }
```
`this.setCurrentDimensions(items)` gives a 500 error. Instead of passing `items`, it should pass `this.projectId`, i.e. `this.setCurrentDimensions(this.projectId)`, as the underlying code will call the API based on the project ID.

artikandri replied:
Hello, thank you for pointing it out, I missed it. Added

teddy-f-47 commented on frontend/services/application/dimension/dimensionData.ts:
```diff
@@ -0,0 +1,23 @@
+    this.id = item.id
+    this.name = item.name
+    this.type = item.type
+    this.metadata = item.metadata 
```
On line 17, there is a TS error `Type 'string' is not assignable to type 'any[]'`. Maybe we need to change line 11 from `metadata: any[]` to `metadata: any` (or alternatively the metadata type in `domain/models/dimension/dimension.ts` needs to be changed to `any[]`). I hope it won't break anything? 😅

artikandri replied:
That's weird, I'm not getting any errors in my `node` tab, but sure.

teddy-f-47 commented on frontend/pages/projects/create.vue:
```diff
@@ -80,9 +89,18 @@ export default Vue.extend({
       })
       delete editedItem.affectiveProjectMode
+
+      if (this.editedItem.projectType === 'DynamicAnnotation') {
+        editedItem.dimension = editedItem.dimension.map((dim: number) => {
```
This probably does not really matter, but I think we can remove the `this.` from line 96 because `editedItem` has been cloned above, just to make it consistent with the code lines above it....

artikandri replied:
Yes, it won't affect the project, but for the sake of consistency, I agree that `this` should be removed. Thank you

teddy-f-47 commented on frontend/i18n/fr/projects/labels.js:
```diff
@@ -17,5 +17,24 @@ export default {
+  dimensionType: 'Type de dimension',
+  min : 'Min',
+  max : 'Max',
+  pas: 'Pas',
```
Line 31 field gets accidentally translated, it should be `step: 'Pas',`

teddy-f-47 commented on frontend/domain/models/project/project.ts:
```diff
@@ -165,15 +172,21 @@ export class ProjectWriteItem {
+      dimension: this.dimension,
     }
+
+    if(result.project_type !== "DynamicAnnotation") {
+      result.tags = this.tags? this.tags.map((tag) => ({ text: tag })) : []
+    }
```
Just wondering, why is this code necessary? Is there some problem with tags in dynamic annotation project type? 😅

artikandri replied:
When I first rebased to the `be` branch, the `tags` property will give error when it's used with the `dynamic annotation` project, so I excluded it and communicated the errors with the BE. It has been fixed, but I forgot to remove and test it again. I will remove this handling and test it again, thank you

teddy-f-47 reviewed (commented):
Thank you for the great work! I have done a quick test and I have seen the code, but I will need some time to do a more thorough test. In general it looks great, there are a just few minor comments. XD

About the limitations you mentioned, I think they are all acceptable:
 1. a. I think the min and max fields for slider only need to be integer. There is no need to make it accept float or something else. Even if there is a use-case requiring float, for example between 0.0 to 1.0, we can just make the slider ranges from 0 to 100 (or from 0 to 1000, depending on how precise it needs to be), and later outside doccano when processing/analyizing the research data the value can be divided by 100 (or 1000).
 1. b. How about hiding the tick labels entirely? For example, a slider for "Zachwyt" with ticks from 0 to 10 does not need to actually show the tick labels 0, 1, 2, ..., 10. It only needs to show the left-side label "Nie wzbudza zachwytu" and the right-side label "Wzbudza zachwyt w bardzo dużym stopniu". And I remember one of the feedbacks mentioning that the users want to see the selected value without having to count the ticks; to fulfill this need, we just show the thumb in the slider. In short, my suggestion is: **hide the tick labels entirely but always show the thumb**.
 2. I think this "feature" to restore the previous value is not very important so let's just leave it as it is. XD
 3. I think the statistics page should work just fine. It actually collects stats from all projects regardless of the type, so even if there is a new project type it won't be a problem. Nevertheless, I will take some time to test it thoroughly.
 4. I think it is ok.

artikandri commented:
Hello, thank you for your feedbacks. I have adjusted some of the codes to follow your comments, but I will need to gather feedbacks from the others about hiding the tickLabels, thank you. Can you check?

haanh764 reviewed (commented):
Hello, thank you for a great work! The code seems oke for me, however I found a possible bug in the annotation view. In the scr recording below, I created a project with some dimensions, including 3 dimensions from Offensive mode (1 scale, 2 checkboxes). The 1st checkbox seem ok, when I select 1 option and then if I wanna change my mind, I deselect it and I can choose other option. But for the 2nd checkbox, when I deselect my option it gave me 404 error when I select the new one (the reason is it tries to find the deleted answer which is already deleted when we deselect option - you can see it send the request to api and got not found message). Could you check that behaviour please?

Besides, I have a question about the checkbox. Previously, when design the backend part, I thought that for Offensive and Humour mode, each checkbox is converted to a dimension. So I imagined that for example when creating the project I choose dimension `W jaki sposób? - Nawoływanie do przemocy` and `W jaki sposób? - Upokorzenie`, in annotation view I will see 2 checkbox corresponding to 2 dimension only, and the user can select them or not. Since I thought that the objective of this project is to reduce the number of dimensions. But as I see here, it is 2 dimension = 2 questions, and the options for each are the whole dimensions (options) for question `W jaki sposób?`, which is completely similar to the affective annotation project :sweat_smile: So I am confuse now, or maybe I understand it incorrectly?


https://user-images.githubusercontent.com/17596118/227322697-830f84b9-4614-4377-ab73-1927796e5e92.mp4

haanh764 commented:
Another comment is for the slider, in the 1st time selecting the slider value, there's a small red text `wymagany` under the slider (as in the video below). I understand it's for warning required dimension, but it's showing when I do my annotation work, not like when I want to confirm without doing annotation :sweat_smile: It makes me a bit uncomfortable, but I don't know is it intentionally(?) :sweat_smile: 

https://user-images.githubusercontent.com/17596118/227325189-509888d1-1af0-4b65-ba3e-26b097dd4997.mp4

artikandri commented:
> Hello, thank you for a great work! The code seems oke for me, however I found a possible bug in the annotation view. In the scr recording below, I created a project with some dimensions, including 3 dimensions from Offensive mode (1 scale, 2 checkboxes). The 1st checkbox seem ok, when I select 1 option and then if I wanna change my mind, I deselect it and I can choose other option. But for the 2nd checkbox, when I deselect my option it gave me 404 error when I select the new one (the reason is it tries to find the deleted answer which is already deleted when we deselect option - you can see it send the request to api and got not found message). Could you check that behaviour please?
> 
> Besides, I have a question about the checkbox. Previously, when design the backend part, I thought that for Offensive and Humour mode, each checkbox is converted to a dimension. So I imagined that for example when creating the project I choose dimension `W jaki sposób? - Nawoływanie do przemocy` and `W jaki sposób? - Upokorzenie`, in annotation view I will see 2 checkbox corresponding to 2 dimension only, and the user can select them or not. Since I thought that the objective of this project is to reduce the number of dimensions. But as I see here, it is 2 dimension = 2 questions, and the options for each are the whole dimensions (options) for question `W jaki sposób?`, which is completely similar to the affective annotation project sweat_smile So I am confuse now, or maybe I understand it incorrectly?
> 
> This is the recording of the behaviour 404 error I describe above https://user-images.githubusercontent.com/17596118/227322697-830f84b9-4614-4377-ab73-1927796e5e92.mp4

Hello, thank you for finding the bugs XD   The previous bug was caused by Vue not updating the component data thoroughly (because the dimensions were saved as array/list and Vue always has trouble in dealing with these kinds of stuffs), so I fixed it by adding $forceUpdate. Likewise, the 3rd bug was caused by the component data being updated even though its still in submit mode, so I fixed it

About the 2nd "bug", it was caused by my misunderstanding about question = dimension, even though it should be option = dimension. Fixing this will take time, but I will give an update when I fixed it.

teddy-f-47 commented on frontend/components/tasks/dynamicAnnotation/DimensionInput.vue:
```diff
@@ -0,0 +1,277 @@
+            <v-chip v-if="index < 3">
+              <span>{{ item.text }}</span>
+            </v-chip>
+            <span v-if="index === 3" class="text-grey text-caption align-self-center">
+              (+{{ selected.length - 3 }} $t('overview.others'))
+            </span>
```
Hello, there is a very minor visual bug here, I think `$t('overview.others')` should be put inside the curly bracket, otherwise it will appear as a string:
<img width="960" alt="Capture" src="https://user-images.githubusercontent.com/64476430/227880906-eeaeb491-059c-4172-af41-3419dd9382e5.PNG">

Not really important but I suppose it can be fixed quickly 😅

artikandri replied:
lol thank you

teddy-f-47 commented:
Unfortunately I found another bug when using a slider with a "nie wiem" checkbox. I'm not sure if it is due to recent changes in the code, but I am unable to mark the checkbox. At first it is marked, but then the slider changes value back to 0.

https://user-images.githubusercontent.com/64476430/227893101-18137258-2378-483e-a09d-b1a5b742b356.mp4

When I monitor the value in the database, when the "nie wiem" checkbox is marked the value is stored as -1, but then the slider changes back to 0 and the value in the database is also updated to 0. So most probably due to the input component in the FE calling the update twice somehow. 😅

artikandri commented:
> Unfortunately I found another bug when using a slider with a "nie wiem" checkbox. I'm not sure if it is due to recent changes in the code, but I am unable to mark the checkbox. At first it is marked, but then the slider changes value back to 0.
> 
> When I monitor the value in the database, when the "nie wiem" checkbox is marked the value is stored as -1, but then the slider changes back to 0 and the value in the database is also updated to 0. So most probably due to the input component in the FE calling the update twice somehow. sweat_smile

Hello, thank you for finding the bug. Yeah, I think it was because of the recent update. I will notify here when I have resolved this.

artikandri commented:
Hello, thank you for the wait. I have fixed the mentioned bugs above and the validations, can you check?

haanh764 commented:
> Hello, thank you for the wait. I have fixed the mentioned bugs above and the validations, can you check?

Hello, thank you for your fix bugs. However, I'm testing it and I found this error
```
  components/tasks/dynamicAnnotation/SliderInput.vue 353:12  root stylesheet
```
Could you check it please :sweat_smile:

artikandri commented:
> > Hello, thank you for the wait. I have fixed the mentioned bugs above and the validations, can you check?
> 
> Hello, thank you for your fix bugs. However, I'm testing it and I found this error
> 
> ```
>   components/tasks/dynamicAnnotation/SliderInput.vue 353:12  root stylesheet
> ```
> 
> Could you check it please sweat_smile

Sure; I think that "error" is logged because I removed the `scoped` part from the `style`, which allowed me to style the `slider`, and probably caused the styling in the component to conflict with the compiled stylesheet somewhere. However, adding this line back will not allow me to style the component, so there will be some slight changes.

teddy-f-47 commented on frontend/components/label/FormCreateDimension.vue:
```diff
@@ -0,0 +1,515 @@
+            readonly: Number(base.formData.readOnly)
+          }
+        ]
+      }
```
I just found another bug here. If I already created custom dynamic dimensions in a previous project, I won't be able to make custom dynamic dimensions in another project, as it always returns 404 error.
<img width="960" alt="ss1" src="https://user-images.githubusercontent.com/64476430/228464018-a4c46185-8da9-4c7c-8813-b98060cce41e.PNG">

The error says `dimension meta data with this codename already exists`. The problem lies on line 340 and 347. Here, the numbering for the codename is based on the length of the assigned dynamic dimensions in the project, whereas the codename should be unique globally. So, for example, if I already have a custom dim with codename `DIM_DYN_1` from a previous project, I can't make new custom dims in a new project if the new project does not have any assigned dynamic dimensions (the length will be 0, so the codename when I try to make a new custom dim will also be `DIM_DYN_1`).

I believe the solution is that, the numbering for the codename should be based on the length of ALL dimensions (instead of just assigned dimensions - line 340). 😅 I'm sorry I didn't catch the bug earlier, previously I tried creating dims repeatedly but within the same project, so the bug didn't happen then. I hope the fix can be pretty quick. 😅

artikandri replied:
Thanks for finding the bug, updated.

teddy-f-47 replied:
Hello, I have tested again and unfortunately it hasn't been fixed. I see that we now use `base.allDimensions`, but this still fails because the items in `base.allDimensions` do not have the `.group` property (I think the `.group` property is only assigned to items in `base.assignedDimensions`).
<img width="631" alt="Capture" src="https://user-images.githubusercontent.com/64476430/228527251-fcec2e5f-050b-4534-a83c-beee6bf3c4e4.PNG">

A quick fix would be adding the following code to line 289 inside `setDimensionList()` - may require refactor because it is copied and adapted from the same mechanism in `setAssignedDimensions()`:
```
      base.allDimensions = base.allDimensions.map((item: any) => {
        const groupMap: any = {
          DIM_OTH: 'Others',
          DIM_OF: 'Offensive',
          DIM_HUM: 'Humor',
          DIM_EMO: 'Emotions'
        }
        if (item.metadata && item.metadata.length) {
          const { codename } = item.metadata[0]
          const groupMapKey: string =
            Object.keys(groupMap).find((key) => codename.includes(key)) || ''
          item.group = groupMap[groupMapKey] || 'Dynamic'
        }
        return item
      })
```

artikandri replied:
I made a new function to assign the group and loaded it here, I hope this should be fixed by now. Please let me know if you still encountered troubles. Thank you.

teddy-f-47 reviewed (commented):
Thank you very much for your hard work. Overall I think the UI already looks great, and I have tested the dim creation, dim assignment, dataset import, dataset export, and the stats pages, they work properly. The slider input component also works well.

There is unfortunately still one bug on the annotation page related with the checkbox input. The sliders work well, but the checkboxes seem to be prone to buggy behavior. Sometimes the checkboxes do not save their data into DB. I recorded the behavior in two videos:

In the first video, I tried to mark two checkboxes under the same question. With the current validation, I expected it to show an error marking, but instead my answers got deleted. After that, the checkboxes do not save their data into DB.
https://user-images.githubusercontent.com/64476430/228498550-562448d9-9122-4287-96a7-b9bb79edf226.mp4

In the second video, I tried to work on other questions, at first some of them do store data into DB, but then they eventually do not save into DB too.
https://user-images.githubusercontent.com/64476430/228498988-5c373db1-ac50-4a7e-a595-3147bbace29b.mp4

I haven't been able to pinpoint the exact cause of the buggy behavior. It seems to be triggered by me trying to mark two checkboxes under the same question and then my answers got deleted, but I am not sure if it is the exact cause, as I had to repeat two or three times before the bug happened. Maybe something with the condition inside `onDynamicComponentRemoveLabel` or `onDynamicComponentAddLabel`.

teddy-f-47 commented:
Additionally, @haanh764, I am also wondering about the validation in the default checkbox dims. Currently the metadata stored in BE sets `'max_answer_number'=1`, which is the reason of the one-answer-only validation on FE for each question with the default dims.

But probably we want to make the max answer equal to the total number of options in each question? Because right now I can't select more than one checkbox for each checkbox question with the default dims

haanh764 commented:
> > > Hello, thank you for the wait. I have fixed the mentioned bugs above and the validations, can you check?
> > 
> > 
> > Hello, thank you for your fix bugs. However, I'm testing it and I found this error
> > ```
> >   components/tasks/dynamicAnnotation/SliderInput.vue 353:12  root stylesheet
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Could you check it please sweat_smile
> 
> Sure; I think that "error" is logged because I removed the `scoped` part from the `style`, which allowed me to style the `slider`, and probably caused the styling in the component to conflict with the compiled stylesheet somewhere. However, adding this line back will not allow me to style the component, so there will be some slight changes.

Hello, I confirm that after I re-compiled as you said it works ok without the scope, so I think you can remove the `scope`. 

Btw, about the bug that checkbox dimensions are not showing. I have tested a bit more, and the behaviour is like this:
- If I choose only few options (dimensions) -> it wont be shown
- As you said, I tried to add more and more options (dimensions), it will show all dimensions belong to that question. For example, for the dimension `Charakter humoru, śmieszy ze względu na - ...`, I added 7/15 options (dimensions), and then it shows me the whole 15 dimensions.
![image](https://user-images.githubusercontent.com/17596118/228501588-ad3534d9-a32e-44a2-8178-83a90f7c056e.png)
![image](https://user-images.githubusercontent.com/17596118/228501694-506b63c8-eab1-4da4-b3bc-28268ddf5004.png)
Just note this here for more detailed info for you :sweat_smile:

haanh764 commented:
> checkbox

Hello, the reason I set it to 1 because previously, we considered each option is a dimension, and with that each option should have only 1 answer (select or not). That's why I set it to 1. But as I see with the current validation method from FE, it should be changed to the numbers of dimensions belonging to the same question. So I will change it no problem :sweat_smile:

haanh764 commented:
> Additionally, @haanh764, I am also wondering about the validation in the default checkbox dims. Currently the metadata stored in BE sets `'max_answer_number'=1`, which is the reason of the one-answer-only validation on FE for each question with the default dims.
> 
> But probably we want to make the max answer equal to the total number of options in each question? Because right now I can't select more than one checkbox for each checkbox question with the default dims

Hello, the reason I set it to 1 because previously, we considered each option is a dimension, and with that each option should have only 1 answer (select or not). That's why I set it to 1. But as I see with the current validation method from FE, it should be changed to the numbers of dimensions belonging to the same question. So I will change it no problem :sweat_smile:

haanh764 commented:
@teddy-f-47 @artikandri already change `max_answer_number` to number of options (dimensions) in each question. You can rebase #94 and migrate the db again

artikandri commented:
> > > > Hello, thank you for the wait. I have fixed the mentioned bugs above and the validations, can you check?
> > > 
> > > 
> > > Hello, thank you for your fix bugs. However, I'm testing it and I found this error
> > > ```
> > >   components/tasks/dynamicAnnotation/SliderInput.vue 353:12  root stylesheet
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Could you check it please sweat_smile
> > 
> > 
> > Sure; I think that "error" is logged because I removed the `scoped` part from the `style`, which allowed me to style the `slider`, and probably caused the styling in the component to conflict with the compiled stylesheet somewhere. However, adding this line back will not allow me to style the component, so there will be some slight changes.
> 
> Hello, I confirm that after I re-compiled as you said it works ok without the scope, so I think you can remove the `scope`.
> 
> Btw, about the bug that checkbox dimensions are not showing. I have tested a bit more, and the behaviour is like this:
> 
> * If I choose only few options (dimensions) -> it wont be shown
> * As you said, I tried to add more and more options (dimensions), it will show all dimensions belong to that question. For example, for the dimension `Charakter humoru, śmieszy ze względu na - ...`, I added 7/15 options (dimensions), and then it shows me the whole 15 dimensions.
>   ![image](https://user-images.githubusercontent.com/17596118/228501588-ad3534d9-a32e-44a2-8178-83a90f7c056e.png)
>   ![image](https://user-images.githubusercontent.com/17596118/228501694-506b63c8-eab1-4da4-b3bc-28268ddf5004.png)
>   Just note this here for more detailed info for you sweat_smile

Hello, thank you for finding the bug. I previously only tested with all dimensions assigned; hence the naive bugs appearing XDXD  It was because I naively assumed that all dimensions to a given original question will be assigned, but of course we have the freedom not to choose.  I have fixed this, can you check?

haanh764 commented:
> > > > > Hello, thank you for the wait. I have fixed the mentioned bugs above and the validations, can you check?
> > > > 
> > > > 
> > > > Hello, thank you for your fix bugs. However, I'm testing it and I found this error
> > > > ```
> > > >   components/tasks/dynamicAnnotation/SliderInput.vue 353:12  root stylesheet
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Could you check it please sweat_smile
> > > 
> > > 
> > > Sure; I think that "error" is logged because I removed the `scoped` part from the `style`, which allowed me to style the `slider`, and probably caused the styling in the component to conflict with the compiled stylesheet somewhere. However, adding this line back will not allow me to style the component, so there will be some slight changes.
> > 
> > 
> > Hello, I confirm that after I re-compiled as you said it works ok without the scope, so I think you can remove the `scope`.
> > Btw, about the bug that checkbox dimensions are not showing. I have tested a bit more, and the behaviour is like this:
> > 
> > * If I choose only few options (dimensions) -> it wont be shown
> > * As you said, I tried to add more and more options (dimensions), it will show all dimensions belong to that question. For example, for the dimension `Charakter humoru, śmieszy ze względu na - ...`, I added 7/15 options (dimensions), and then it shows me the whole 15 dimensions.
> >   ![image](https://user-images.githubusercontent.com/17596118/228501588-ad3534d9-a32e-44a2-8178-83a90f7c056e.png)
> >   ![image](https://user-images.githubusercontent.com/17596118/228501694-506b63c8-eab1-4da4-b3bc-28268ddf5004.png)
> >   Just note this here for more detailed info for you sweat_smile
> 
> Hello, thank you for finding the bug. I previously only tested with all dimensions assigned; hence the naive bugs appearing XDXD It was because I naively assumed that all dimensions to a given original question will be assigned, but of course we have the freedom not to choose. I have fixed this, can you check?

Hello, thank you for your quick fix but unfortunately after recompiling I still have that bug, her's the dimensions I selected and the annotation view:
![image](https://user-images.githubusercontent.com/17596118/228514347-0022948f-b4c0-400e-99e2-099ccc90efc7.png)
![image](https://user-images.githubusercontent.com/17596118/228514389-a07346a4-05ea-43d8-96fa-7420152da239.png)
Could you check it again pls?

artikandri commented:
> > > > > > Hello, thank you for the wait. I have fixed the mentioned bugs above and the validations, can you check?
> > > > > 
> > > > > 
> > > > > Hello, thank you for your fix bugs. However, I'm testing it and I found this error
> > > > > ```
> > > > >   components/tasks/dynamicAnnotation/SliderInput.vue 353:12  root stylesheet
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > Could you check it please sweat_smile
> > > > 
> > > > 
> > > > Sure; I think that "error" is logged because I removed the `scoped` part from the `style`, which allowed me to style the `slider`, and probably caused the styling in the component to conflict with the compiled stylesheet somewhere. However, adding this line back will not allow me to style the component, so there will be some slight changes.
> > > 
> > > 
> > > Hello, I confirm that after I re-compiled as you said it works ok without the scope, so I think you can remove the `scope`.
> > > Btw, about the bug that checkbox dimensions are not showing. I have tested a bit more, and the behaviour is like this:
> > > 
> > > * If I choose only few options (dimensions) -> it wont be shown
> > > * As you said, I tried to add more and more options (dimensions), it will show all dimensions belong to that question. For example, for the dimension `Charakter humoru, śmieszy ze względu na - ...`, I added 7/15 options (dimensions), and then it shows me the whole 15 dimensions.
> > >   ![image](https://user-images.githubusercontent.com/17596118/228501588-ad3534d9-a32e-44a2-8178-83a90f7c056e.png)
> > >   ![image](https://user-images.githubusercontent.com/17596118/228501694-506b63c8-eab1-4da4-b3bc-28268ddf5004.png)
> > >   Just note this here for more detailed info for you sweat_smile
> > 
> > 
> > Hello, thank you for finding the bug. I previously only tested with all dimensions assigned; hence the naive bugs appearing XDXD It was because I naively assumed that all dimensions to a given original question will be assigned, but of course we have the freedom not to choose. I have fixed this, can you check?
> 
> Hello, thank you for your quick fix but unfortunately after recompiling I still have that bug, her's the dimensions I selected and the annotation view: ![image](https://user-images.githubusercontent.com/17596118/228514347-0022948f-b4c0-400e-99e2-099ccc90efc7.png) ![image](https://user-images.githubusercontent.com/17596118/228514389-a07346a4-05ea-43d8-96fa-7420152da239.png) Could you check it again pls?

Fixed; I also fixed the preview in which I excluded the options that had yet to be added, can you check?

teddy-f-47 commented:
Hi, I am adding a bit more information about the bug I described earlier about checkboxes not saving into DB (https://github.com/CLARIN-PL/doccano/pull/97#pullrequestreview-1362721978).

I tried adding some console logs inside the `onDynamicComponentUpdateScale`, `onDynamicComponentRemoveLabel` and `onDynamicComponentAddLabel` as you can see in the following recording. Then I tried filling the checkboxes. At first they worked (they saved data into DB, indicated by the logs into the console and the "submitting" animation of temporary graying of the component). But then they suddenly stopped working (the checkboxes still appear ok on UI, but they don't save data into DB, indicated by no logs printed in console and no "submitting" animation). It is really weird 😅 

https://user-images.githubusercontent.com/64476430/228538404-432314d9-7c45-4caf-ba40-5ed4b66fcfd2.mp4

I also noticed that the bug happened after I deselected an option. I suspect this is somewhat related with the bug that Anh's previously mentioned:
> Hello, thank you for a great work! The code seems oke for me, however I found a possible bug in the annotation view. In the scr recording below, I created a project with some dimensions, including 3 dimensions from Offensive mode (1 scale, 2 checkboxes). The 1st checkbox seem ok, when I select 1 option and then if I wanna change my mind, I deselect it and I can choose other option. But for the 2nd checkbox, when I deselect my option it gave me 404 error when I select the new one (the reason is it tries to find the deleted answer which is already deleted when we deselect option - you can see it send the request to api and got not found message). Could you check that behaviour please?

The difference is that it no longer shows 404 error, but it doesn't save to DB.

haanh764 commented:
> > > > > > > Hello, thank you for the wait. I have fixed the mentioned bugs above and the validations, can you check?
> > > > > > 
> > > > > > 
> > > > > > Hello, thank you for your fix bugs. However, I'm testing it and I found this error
> > > > > > ```
> > > > > >   components/tasks/dynamicAnnotation/SliderInput.vue 353:12  root stylesheet
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > >     
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >     
> > > > > >   
> > > > > > Could you check it please sweat_smile
> > > > > 
> > > > > 
> > > > > Sure; I think that "error" is logged because I removed the `scoped` part from the `style`, which allowed me to style the `slider`, and probably caused the styling in the component to conflict with the compiled stylesheet somewhere. However, adding this line back will not allow me to style the component, so there will be some slight changes.
> > > > 
> > > > 
> > > > Hello, I confirm that after I re-compiled as you said it works ok without the scope, so I think you can remove the `scope`.
> > > > Btw, about the bug that checkbox dimensions are not showing. I have tested a bit more, and the behaviour is like this:
> > > > 
> > > > * If I choose only few options (dimensions) -> it wont be shown
> > > > * As you said, I tried to add more and more options (dimensions), it will show all dimensions belong to that question. For example, for the dimension `Charakter humoru, śmieszy ze względu na - ...`, I added 7/15 options (dimensions), and then it shows me the whole 15 dimensions.
> > > >   ![image](https://user-images.githubusercontent.com/17596118/228501588-ad3534d9-a32e-44a2-8178-83a90f7c056e.png)
> > > >   ![image](https://user-images.githubusercontent.com/17596118/228501694-506b63c8-eab1-4da4-b3bc-28268ddf5004.png)
> > > >   Just note this here for more detailed info for you sweat_smile
> > > 
> > > 
> > > Hello, thank you for finding the bug. I previously only tested with all dimensions assigned; hence the naive bugs appearing XDXD It was because I naively assumed that all dimensions to a given original question will be assigned, but of course we have the freedom not to choose. I have fixed this, can you check?
> > 
> > 
> > Hello, thank you for your quick fix but unfortunately after recompiling I still have that bug, her's the dimensions I selected and the annotation view: ![image](https://user-images.githubusercontent.com/17596118/228514347-0022948f-b4c0-400e-99e2-099ccc90efc7.png) ![image](https://user-images.githubusercontent.com/17596118/228514389-a07346a4-05ea-43d8-96fa-7420152da239.png) Could you check it again pls?
> 
> Fixed; I also fixed the preview in which I excluded the options that had yet to be added, can you check?

Hello, I am testing it again but now it gives me `404 error` on both annotation view and also when I'm trying to create a new dynamic annotation project :sweat_smile:  Could you check it again?
I guess the error could come from your recent fix here:
![Screenshot from 2023-03-29 17-31-59](https://user-images.githubusercontent.com/17596118/228590724-49a366a5-7d1c-405a-a56e-8572b4287262.png)

artikandri commented:
Sorry for the many bugs, I should have switched my job to a bug farmer instead, lol. 

Jokes aside, I hope I have fixed the checkbox bugs. I also don't really understand what caused it, but I added isSubmitting flag just in case. I think the bug happened only once on first load, which means that some of the formData dimensions data are not set correctly, but I haven't found the exact pattern to reproduce it, so I can't make sure. 

Also those yarn errors are caused by me forgetting to fix the `typescript errors`, my bad. Fixed.

haanh764 commented:
Hello again :sweat_smile: the checkboxes bug I mentioned before and the 404 error have gone, thank you for the fix. However, additional information for you with the bug Teddy mentioned above is here since Im having the same bug. As in the video, I have selected the option for the 1st and 2nd questions, but it didn't trigger the api to write the label, that's why there is nothing to be saved in database. But it's weird that when I select an option for the 3rd question, suddenly it triggered the api and somehow remove selected in the 2 previous questions :sweat_smile: 

https://user-images.githubusercontent.com/17596118/228627451-636cada3-755e-4050-a86b-612e9bd230b0.mp4

haanh764 commented:
> Sorry for the many bugs, I should have switched my job to a bug farmer instead, lol.
> 
> Jokes aside, I hope I have fixed the checkbox bugs. I also don't really understand what caused it, but I added isSubmitting flag just in case. I think the bug happened only once on first load, which means that some of the formData dimensions data are not set correctly, but I haven't found the exact pattern to reproduce it, so I can't make sure.
> 
> Also those yarn errors are caused by me forgetting to fix the `typescript errors`, my bad. Fixed.

Hello, I will test again with your fix. But I can confirm that it doesn't happen only for the first load. As you can see in the video in my comment above, it's already the 2nd or 3rd example :sweat_smile:

artikandri commented:
> > Sorry for the many bugs, I should have switched my job to a bug farmer instead, lol.
> > Jokes aside, I hope I have fixed the checkbox bugs. I also don't really understand what caused it, but I added isSubmitting flag just in case. I think the bug happened only once on first load, which means that some of the formData dimensions data are not set correctly, but I haven't found the exact pattern to reproduce it, so I can't make sure.
> > Also those yarn errors are caused by me forgetting to fix the `typescript errors`, my bad. Fixed.
> 
> Hello, I will test again with your fix. But I can confirm that it doesn't happen only for the first load. As you can see in the video in my comment above, it's already the 2nd or 3rd example sweat_smile

Yes, I can see that. However I cannot seem to reproduce it really well even though I have even slowed my localhost internet connection to 100ms, so I will wait for your test on my recent fix, sorry :crying_cat_face:

haanh764 commented:
> > > Sorry for the many bugs, I should have switched my job to a bug farmer instead, lol.
> > > Jokes aside, I hope I have fixed the checkbox bugs. I also don't really understand what caused it, but I added isSubmitting flag just in case. I think the bug happened only once on first load, which means that some of the formData dimensions data are not set correctly, but I haven't found the exact pattern to reproduce it, so I can't make sure.
> > > Also those yarn errors are caused by me forgetting to fix the `typescript errors`, my bad. Fixed.
> > 
> > 
> > Hello, I will test again with your fix. But I can confirm that it doesn't happen only for the first load. As you can see in the video in my comment above, it's already the 2nd or 3rd example sweat_smile
> 
> Yes, I can see that. However I cannot seem to reproduce it really well even though I have even slowed my localhost internet connection to 100ms, so I will wait for your test on my recent fix, sorry crying_cat_face

Hello it's me again :sweat_smile: It seems that the root of this bug is not from the slow internet which leads to async bug, it's from where I click to select the option. After your recent changes, I can realize that if I select the option by clicking into the box precisely, then it triggers the api -> save label to database. But if I click to the text -> it becomes gray and no api trigger -> nothing saved. You can see in the picture 2 different results.

![image](https://user-images.githubusercontent.com/17596118/228635937-3b409239-d581-4a7c-9438-991deaeb7135.png)

haanh764 commented:
Here's the behaviour, if I click to the text it blocked the whole dimension. And a weird behaviour is when I select another dimension in other question, it released the blocked dimensions, which explains the behaviour I describe above :sweat_smile: 

https://user-images.githubusercontent.com/17596118/228636240-fdd2b635-b8c5-4aae-bffb-4a1c837eff87.mp4

artikandri commented:
> Here's the behaviour, if I click to the text it blocked the whole dimension. And a weird behaviour is when I select another dimension in other question, it released the blocked dimensions, which explains the behaviour I describe above sweat_smile

Hello, thank you for finding the cause. It was my fault for not handling the click event on the label; I only handled the click on the box instead. I have fixed this bug, can you check?

haanh764 commented:
> > Here's the behaviour, if I click to the text it blocked the whole dimension. And a weird behaviour is when I select another dimension in other question, it released the blocked dimensions, which explains the behaviour I describe above sweat_smile
> 
> Hello, thank you for finding the cause. It was my fault for not handling the click event on the label; I only handled the click on the box instead. I have fixed this bug, can you check?

Thank you for your fix. I have tested again and that bug has been solved finally :sweat_smile:

teddy-f-47 commented on frontend/components/tasks/dynamicAnnotation/DimensionInput.vue:
```diff
@@ -0,0 +1,287 @@
+          value: item.id,
+          text: item.name,
+          type: item.type
+        }
```
The `metadata` field is missing here, resulting in the dims having an `undefined` group. 😅 Adding `metadata: item.metadata` solves the issue.
<img width="960" alt="Capture" src="https://user-images.githubusercontent.com/64476430/228772757-3116ffad-c3b2-40d7-b3be-d7877504bd94.PNG">

Notably the issue only happens with the DimensionInput component, the annotation page is fine. The dims on the annotation page are grouped properly. :)

artikandri replied:
Thanks, I have fixed this bug, can you check?

teddy-f-47 reviewed (commented):
Once again, thank you very much for the hard work! It was a very good catch by Anh about the root cause of the checkbox bug, I didn't notice it. XD I have tested everything once more, and on my side everything seems to be working well now. There is still one very minor comment, but I think I am ready to give my approval after the comment is solved. 😅

haanh764 commented:
Hello, I think there's another bug with the slider dimension with checkbox (the 'Nie wiem'....). As we discussed earlier they won't be shown standalone and it's ok, but with the slider dimension with checkbox it should be shown right? For eg here, you can see that I select dimension `Zrozumialy` which should have the checkbox `Nie wiem` (DIM_OTH10), but it is not shown here

![image](https://user-images.githubusercontent.com/17596118/228792205-4a89ae69-9530-420a-8600-cc4f7e59b2ff.png)

Another problem is when I create a new slider dimension with checkbox, in the selection for checkbox codename, there's no codename shown for choosing, and I can't choose anything. I suppose that here we should have the option of `DIM_OTH10` or `DIM_OTH11` (`Nie wiem`, or `Nie mam zyczen`)

![image](https://user-images.githubusercontent.com/17596118/228793067-79b60136-6b60-4d7e-b2f7-da8d02b2d748.png)

teddy-f-47 commented:
> Hello, I think there's another bug with the slider dimension with checkbox (the 'Nie wiem'....). As we discussed earlier they won't be shown standalone and it's ok, but with the slider dimension with checkbox it should be shown right? For eg here, you can see that I select dimension `Zrozumialy` which should have the checkbox `Nie wiem` (DIM_OTH10), but it is not shown here
> 
> ![image](https://user-images.githubusercontent.com/17596118/228792205-4a89ae69-9530-420a-8600-cc4f7e59b2ff.png)
> 
> Another problem is when I create a new slider dimension with checkbox, in the selection for checkbox codename, there's no codename shown for choosing, and I can't choose anything.
> 
> ![image](https://user-images.githubusercontent.com/17596118/228793067-79b60136-6b60-4d7e-b2f7-da8d02b2d748.png)

for this case i simply created a new custom checkbox manually and named it "nie wiem", and then i was able to attach it to a new slider. imo it is usable this way

haanh764 commented:
> > Hello, I think there's another bug with the slider dimension with checkbox (the 'Nie wiem'....). As we discussed earlier they won't be shown standalone and it's ok, but with the slider dimension with checkbox it should be shown right? For eg here, you can see that I select dimension `Zrozumialy` which should have the checkbox `Nie wiem` (DIM_OTH10), but it is not shown here
> > ![image](https://user-images.githubusercontent.com/17596118/228792205-4a89ae69-9530-420a-8600-cc4f7e59b2ff.png)
> > Another problem is when I create a new slider dimension with checkbox, in the selection for checkbox codename, there's no codename shown for choosing, and I can't choose anything.
> > ![image](https://user-images.githubusercontent.com/17596118/228793067-79b60136-6b60-4d7e-b2f7-da8d02b2d748.png)
> 
> for this case i simply created a new custom checkbox manually and named it "nie wiem", and then i was able to attach it to a new slider. imo it is usable this way

I used the same way as you but there's already that dimension in the database. So it is just user's workaround :sweat_smile: And with this workaround, I just found another bug about the validation of FE that because in database we already have `Nie wiem` dimension, and I added the unique constraint on dimension name (requested from FE) -> then if I create new dimension with name `Nie wiem` -> 404 error :sweat_smile: the FE validation only check the label, not the dimension name in that case

artikandri commented:
Lol, the first `undefined` bugs came from my recent update to use a
centralized function to map the group, sorry for that.

As for Anh's bugs, I think it's because I loaded the dimensions for that
field from the assigned Dimensions and not all Dimensions, but I will check
in a moment

czw., 30 mar 2023, 11:34 użytkownik haanh764 ***@***.***>
napisał:

> Hello, I think there's another bug with the slider dimension with checkbox
> (the 'Nie wiem'....). As we discussed earlier they won't be shown
> standalone and it's ok, but with the slider dimension with checkbox it
> should be shown right? For eg here, you can see that I select dimension
> Zrozumialy which should have the checkbox Nie wiem (DIM_OTH10), but it is
> not shown here
> [image: image]
> <https://user-images.githubusercontent.com/17596118/228792205-4a89ae69-9530-420a-8600-cc4f7e59b2ff.png>
> Another problem is when I create a new slider dimension with checkbox, in
> the selection for checkbox codename, there's no codename shown for
> choosing, and I can't choose anything.
> [image: image]
> <https://user-images.githubusercontent.com/17596118/228793067-79b60136-6b60-4d7e-b2f7-da8d02b2d748.png>
>
> for this case i simply created a new custom checkbox manually and named it
> "nie wiem", and then i was able to attach it to a new slider. imo it is
> usable this way
>
> I used the same way as you but there's already that dimension in the
> database. So it is just user's workaround 😅
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CLARIN-PL/doccano/pull/97#issuecomment-1489994038>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC7U67C7VUHUXU7KSL5JJ3LW6VHSVANCNFSM6AAAAAAWCU6HJM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

haanh764 commented:
Here's the bug with validation I mentioned above :sweat_smile: sorry for being bug finder :sweat_smile: 

https://user-images.githubusercontent.com/17596118/228796332-b4c477fa-a1ee-4bcc-b872-c8bef65473e0.mp4

artikandri commented:
And since I'm a bug farmer we are a good match lol. XD

Just kidding, thank you for catching the bug, I will check and report here
when I have solved it.

czw., 30 mar 2023, 11:42 użytkownik haanh764 ***@***.***>
napisał:

> Here's the bug with validation I mentioned above 😅 sorry for being bug
> finder 😅
>
>
> https://user-images.githubusercontent.com/17596118/228796332-b4c477fa-a1ee-4bcc-b872-c8bef65473e0.mp4
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CLARIN-PL/doccano/pull/97#issuecomment-1490005084>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC7U67DDUO7QREMWW7PVI23W6VIRLANCNFSM6AAAAAAWCU6HJM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

artikandri commented:
> Here's the bug with validation I mentioned above sweat_smile sorry for being bug finder sweat_smile

Hello, thank you for the wait. I think I have fixed this, can you check?

haanh764 commented:
> > Here's the bug with validation I mentioned above sweat_smile sorry for being bug finder sweat_smile
> 
> Hello, thank you for the wait. I think I have fixed this, can you check?

Thank you for the quick fix, the validation is ok now, and the option selection for `with_checkbox` also oke now. But the 1st bug I mentioned above still hasn't been solved unfortunately. These dimensions should come with a checkbox but they're not showing now
![image](https://user-images.githubusercontent.com/17596118/228805998-c93a52b9-e0fa-4fc1-b40b-617ff4b3d1be.png)

Also with the new slider I created, I choose `with_checkbox` so it should come with checkbox `Nie wiem`
![image](https://user-images.githubusercontent.com/17596118/228806268-0de0532f-b0f9-49a1-b278-bf688324ceb1.png)

artikandri commented:
Ah, I see what you meant now. I think it was because the `nie wiem` dimension was not added (because I hide it from the `dimension selection`), so the form couldnt find the dimension in the `assigned` dimensions, my bad. I will fix it by loading all dimensions instead. I will fix it and notify here later.

haanh764 commented:
Hello, thank you so much for your fix. I have tested and it works well now. Great work :clap: 
Btw, there is a pretty minor error in yarn, but it's quite minor so I think I am ready to give my approval after this has been solve 
```
 ERROR  ERROR in pages/projects/create.vue:119:13                                                        13:38:13
TS7034: Variable 'adtDimensions' implicitly has type 'any[]' in some locations where its type cannot be determined.
    117 |     },
    118 |     getAdditionalDimensions(dimensions: number[]) {
  > 119 |       const adtDimensions = []
        |             ^^^^^^^^^^^^^
    120 |       const dimensionsWithAdtDimensions = this.dimensions.filter((dim: any) => {
    121 |         if (dim && dim.metadata.length) {
    122 |           const { config } = dim.metadata[0]

ERROR in pages/projects/create.vue:142:14
TS7005: Variable 'adtDimensions' implicitly has an 'any[]' type.
    140 |         }
    141 |       })
  > 142 |       return adtDimensions
        |              ^^^^^^^^^^^^^
    143 |     },
    144 |     async getBlobScaleData(project: ProjectDTO): Promise<Blob[]> {
    145 |       let fdata = [] as Blob[]
```

artikandri commented:
> Hello, thank you so much for your fix. I have tested and it works well now. Great work clap Btw, there is a pretty minor error in yarn, but it's quite minor so I think I am ready to give my approval after this has been solve
> 
> ```
>  ERROR  ERROR in pages/projects/create.vue:119:13                                                        13:38:13
> TS7034: Variable 'adtDimensions' implicitly has type 'any[]' in some locations where its type cannot be determined.
>     117 |     },
>     118 |     getAdditionalDimensions(dimensions: number[]) {
>   > 119 |       const adtDimensions = []
>         |             ^^^^^^^^^^^^^
>     120 |       const dimensionsWithAdtDimensions = this.dimensions.filter((dim: any) => {
>     121 |         if (dim && dim.metadata.length) {
>     122 |           const { config } = dim.metadata[0]
> 
> ERROR in pages/projects/create.vue:142:14
> TS7005: Variable 'adtDimensions' implicitly has an 'any[]' type.
>     140 |         }
>     141 |       })
>   > 142 |       return adtDimensions
>         |              ^^^^^^^^^^^^^
>     143 |     },
>     144 |     async getBlobScaleData(project: ProjectDTO): Promise<Blob[]> {
>     145 |       let fdata = [] as Blob[]
> ```

Sorry, I always forgot to check `yarn` tab lol. Fixed.

haanh764 commented:
Sorry, I just found another bug due to the recent fix for `Nie wiem` checbox. So the problem is when we select many slider dimension with checkbox (for example the dimensions in `Others` group), then the checkbox dimension (dimension with `id=23` in database) will be duplicated in the message FE send to BE in project creation. And it violates the unique constraint between project and assigned dimensions :sweat_smile: So is it possible to remove the duplication of that dimension? or should I remove the constraint? Ofc, I can remove constraint but it's a bit unsafe

artikandri commented:
> Sorry, I just found another bug due to the recent fix for `Nie wiem` checbox. So the problem is when we select many slider dimension with checkbox (for example the dimensions in `Others` group), then the checkbox dimension (dimension with `id=23` in database) will be duplicated in the message FE send to BE in project creation. And it violates the unique constraint between project and assigned dimensions sweat_smile So is it possible to remove the duplication of that dimension? or should I remove the constraint? Ofc, I can remove constraint but it's a bit unsafe

Sorry for the trouble, I added a function to make it `unique` from FE, can you check?

haanh764 reviewed (approved):
Hello, thank you for your great work :smile: I have tested again and everything works well now. I will give my approval now :smile:

teddy-f-47 reviewed (approved):
Thank you for the fixes, approved! :)
